File: Karato_clinopyroxene_conductivity.m
% ------------------------------------------------------------------------
% FUNCTION: Karato_orthopyroxene_conductivity
% ------------------------------------------------------------------------
% DESCRIPTION:
% Compute the conductivity of Orthopyroxene using Karato's group high temperature
% and pressure experiments.
%
% FORMULA:
% The conductivity is computed using the formula:
%   sigma = A1*Cw^{r1}*exp(-(E1+PV1)/(RT)) + A2*Cw^{r2}*exp(-(E2+PV2)/(RT))
% where:
%   - A1, A2 are the pre-exponential terms which already contain the influence
%     of the Mg# and oxygen fugacity.
%   - Cw is the water content.
%   - r1, r2 are constants that depend on the mechanism of electrical conduction.
%   - E1, E2 are the activation energies.
%   - V1, V2 are the activation volumes.
%   - Subscript 1 stands for polaron conduction.
%   - Subscript 2 stands for proton conduction (e.g., H).
%
% REFERENCES:
% Values of A, r, E, V are from:
%   Karato, S. I. (2011). Water distribution across the mantle transition zone
%   and its implications for global material circulation. Earth and Planetary
%   Science Letters, 301(3–4), 413–423. https://doi.org/10.1016/j.epsl.2010.11.038
%
% UNITS:
%   - temperature: Kelvin (K); 0 K is equivalent to -273.15°C.
%   - water_content: dimensionless, typically in the range 0-1.
%   - pressure: Pascal (Pa); 1 Pa = 1 J/m³ = 1e-6 J/cc.
%
% INPUTS:
%   temperature   - Temperature in Kelvin.
%   water_content - Water content (dimensionless).
%   pressure      - Pressure in Pascal.
%
% OUTPUT:
%   sigma - Electrical conductivity (S/m).
% ------------------------------------------------------------------------

function sigma = Karato_clinopyroxene_conductivity(temperature, water_content, pressure)
    % Constants
    R = 8.314;       % Gas constant, J/(K⋅mol)
    A1 = 10^(2.7);   % Pre-exponential factor for polaron conduction, S/m
    A2 = 10^(2.6);   % Pre-exponential factor for proton conduction, S/m
    r1 = 0;          % Empirical constant for polaron conduction
    r2 = 0.62;       % Empirical constant for proton conduction
    E1 = 147000;     % Activation energy for polaron conduction, J/mol
    E2 = 82000;      % Activation energy for proton conduction, J/mol
    V1 = 0.0;        % Activation volume for polaron conduction, cc/mol
    V2 = 0.0;        % Activation volume for proton conduction, cc/mol

    % Check if temperature is greater than zero
    if any(temperature <= 0)
        error('Temperature must be greater than zero Kelvin.');
    end

    % Ensure pressure is numeric
    if iscell(pressure)
        pressure = cell2mat(pressure); % Convert cell array to numeric if necessary
    end

    % Check if pressure is non-negative
    if any(pressure < 0)
        error('Pressure must be non-negative.');
    end

    % Convert pressure from Pa to J/cc
    pressure = 1e3 * pressure;  % 1 Pa = 1 J/m³ = 1e-6 J/cc

    % Ensure water_content is numeric
    if iscell(water_content)
        water_content = cell2mat(water_content); % Convert cell array to numeric if necessary
    end

    % Check if water_content is within the range of 0 to 1
    if any(water_content < 0) || any(water_content > 1)
        error('Water content must be in the range of 0 to 1.');
    end

    % Calculate conductivities
    sigma_polaron = A1 * water_content.^r1 .* exp(-(E1 + pressure * V1) / (R * temperature));
    sigma_proton =  A2 * water_content.^r2 .* exp(-(E2 + pressure * V2) / (R * temperature));

    % Sum the conductivities
    sigma = sigma_polaron + sigma_proton;
end

File: Karato_clinopyroxene_conductivity.cpp
#include <cmath>
#include <stdexcept>
#include <iostream>

// Function to calculate conductivity
double Karato_orthopyroxene_conductivity(double temperature, double water_content, double pressure) {
    // Constants
    const double R = 8.314;       // Gas constant, J/(K⋅mol)
    const double A1 = std::pow(10, 2.7);   // Pre-exponential factor for polaron conduction, S/m
    const double A2 = std::pow(10, 2.6);   // Pre-exponential factor for proton conduction, S/m
    const double r1 = 0.0;        // Empirical constant for polaron conduction
    const double r2 = 0.62;       // Empirical constant for proton conduction
    const double E1 = 147000;     // Activation energy for polaron conduction, J/mol
    const double E2 = 82000;      // Activation energy for proton conduction, J/mol
    const double V1 = 0.0;        // Activation volume for polaron conduction, cc/mol
    const double V2 = 0.0;        // Activation volume for proton conduction, cc/mol

    // Check if temperature is greater than zero
    if (temperature <= 0) {
        throw std::invalid_argument("Temperature must be greater than zero Kelvin.");
    }

    // Check if pressure is non-negative
    if (pressure < 0) {
        throw std::invalid_argument("Pressure must be non-negative.");
    }

    // Convert pressure from Pa to J/cc (1 Pa = 1 J/m³ = 1e-6 J/cc)
    pressure *= 1e3;

    // Check if water_content is within the range of 0 to 1
    if (water_content < 0 || water_content > 1) {
        throw std::invalid_argument("Water content must be in the range of 0 to 1.");
    }

    // Calculate conductivities
    double sigma_polaron = A1 * std::pow(water_content, r1) * std::exp(-(E1 + pressure * V1) / (R * temperature));
    double sigma_proton =  A2 * std::pow(water_content, r2) * std::exp(-(E2 + pressure * V2) / (R * temperature));

    // Sum the conductivities
    double sigma = sigma_polaron + sigma_proton;

    return sigma;
}

// Example usage
int main() {
    double temperature = 1000.0;  // Example temperature in Kelvin
    double water_content = 0.1;   // Example water content
    double pressure = 1e5;        // Example pressure in Pascal

    try {
        double conductivity = Karato_orthopyroxene_conductivity(temperature, water_content, pressure);
        std::cout << "Conductivity: " << conductivity << " S/m" << std::endl;
    } catch (const std::invalid_argument& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}
